synapse_spamcheck_badlist/BadListFilter.py

```diff
@@ -0,0 +1,212 @@
+        if self._can_we_check_md5 is not None:
+            return self._can_we_check_md5
+        if self._md5_table is None:
+            self._can_we_check_md5 = False
+            return False
+        try:
+            def interaction(db):
+                db.execute("SELECT md5 FROM %s LIMIT 1" % self._md5_table)
+            await self._api.run_db_interaction("Check whether we can check md5", interaction)
+            self._can_we_check_md5 = True
+            logger.info("We can check md5!")
+        except:
+            logger.warn("We CANNOT check md5!")
+            self._can_we_check_md5 = False
+        return self._can_we_check_md5
+
+    async def check_event_for_spam(self, event) -> bool:
+        if event["type"] != "m.room.message":
+            # We only filter messages.
+            return False
+
+        # If it's a text content, look for links.
+        if event["content"]["msgtype"] == "m.text":
+            if not await self.can_we_check_links():
+                return False
+
+            # Check for links in text, both unformatted and formatted.
+            #
+            # We always lower-case the text, as the IWF database is lowercase.
+            for text in [event.get("content", {}).get("body", ""), event.get("content", {}).get("formatted_body", "")]:
+                for match in self._linkifier.match(text.lower()) or []:
+                    link = re.sub(self._scheme_re, "", match.url)
+                    is_bad_link = await self._api.run_db_interaction("Check link against evil db", _db_is_bad_link, self._links_table, link)
+                    if is_bad_link:
+                        logger.info("Rejected bad link")
+                        return True
+
+        # If it's a file, download content, extract hash.
+        elif event["content"]["msgtype"] == "m.file":
+            if not await self.can_we_check_md5():
+                return False
+
+            match = self._mxc_re.match(event["content"]["url"])
+            server_name = None
+            media_id = None
+            if match != None:
+                server_name = match.group('server_name')
+                media_id = match.group('media_id')
+                response = None
+                try:
+                    url = "%s/_matrix/media/r0/download/%s/%s" % (
+                            self._base_url,
+                            urlquote(server_name),
+                            urlquote(media_id)
+                    )
+                    response = await self._api.http_client.request("GET", url)
+                    # FIXME: What do we do if we were rate-limited?
+                except Exception as e:
+                    # In case of timeout or error, there's nothing we can do.
+                    # Let's not take the risk of blocking valid contents.
+                    logger.warn("Could not download media: %s" % e)
+                md5 = hashlib.md5()
+                await response.collect(lambda batch: md5.update(batch))
```

Comment:
I'm not really sure what this is doing -- `IResponse` / `Response` doesn't have a `collect` method?

Anyway, I'm hoping we can iteratively md5-hash things as you're suggesting here. 👍

Reply:
But this is actually a `treq.response`, which does :)
https://treq.readthedocs.io/en/release-20.3.0/api.html#treq.response._Response.collect

Reply:
Interesting. I suspect we should update the type signature in Synapse then.